lib/util/state.js

```diff
@@ -53,62 +56,133 @@ export function getActiveError (otpState) {
   if (errorResponse && errorResponse.error) return errorResponse.error
 }
 
+/**
+ * Helper to get the active search's non-realtime response
+ */
+function getActiveSearchNonRealtimeResponse (otpState) {
+  const search = getActiveSearch(otpState)
+  return search && search.nonRealtimeResponse
+}
+
+/**
+ * Helper to get the active search's realtime response
+ */
+function getActiveSearchRealtimeResponse (otpState) {
+  const search = getActiveSearch(otpState)
+  return search && search.response
+}
+
+// Ignore certain keys that could add significant calculation time to hashing.
+// The alerts are irrelevant, but the intermediateStops, interStopGeometry and
+// steps could have the legGeometry substitute as an equivalent hash value
+const blackListedKeys = [
+  'alerts',
+  'intermediateStops',
+  'interStopGeometry',
+  'steps'
+]
+// make blackListedKeys into an object due to superior lookup performance
+const blackListedKeyLookup = {}
+blackListedKeys.forEach(key => {
+  blackListedKeyLookup[key] = true
+})
+
+/**
+ * A memoized function to hash the itinerary.
+ * NOTE: It can take a while (>30ms) for the object-hash library to calculate
+ * an itinerary's hash for some lengthy itineraries. If better performance is
+ * desired, additional values to blackListedKeys should be added to avoid
+ * spending extra time hashing values that wouldn't result in different
+ * itineraries.
+ */
+const hashItinerary = memoize(
+  itinerary => hash(
+    itinerary,
+    { excludeKeys: key => blackListedKeyLookup[key] }
+  )
+)
+
 /**
  * Get the active itineraries for the active search, which is dependent on
  * whether realtime or non-realtime results should be displayed
  * @param {Object} otpState the OTP state object
  * @return {Array}      array of itinerary objects from the OTP plan response,
  *                      or null if there is no active search
  */
-export function getActiveItineraries (otpState) {
-  const search = getActiveSearch(otpState)
-  const { config, useRealtime } = otpState
-  // set response to use depending on useRealtime
-  const response = !search
-    ? null
-    : useRealtime
-      ? search.response
-      : search.nonRealtimeResponse
-  const itineraries = []
-  if (response) {
-    response.forEach(res => {
-      if (res && res.plan) itineraries.push(...res.plan.itineraries)
-    })
-  }
-  const {filter, sort} = otpState.filter
-  const {direction, type} = sort
-  const filteredItineraries = itineraries
-    .filter(itinerary => {
-      switch (filter) {
-        case 'ALL':
-          return true
-        case 'ACTIVE':
-          // ACTIVE filter checks that the only modes contained in the itinerary
-          // are 'active' (i.e., walk or bike). FIXME: add micromobility?
-          let allActiveModes = true
-          itinerary.legs.forEach(leg => {
-            if (!isWalk(leg.mode) && !hasBike(leg.mode)) allActiveModes = false
+export const getActiveItineraries = createSelector(
+  otpState => otpState.config,
```

Comment:
How is `otpState.config` relevant to the selector? (It doesn't change within a session, does it?)

Reply:
It doesn't change, but it's used in the `sortItineraries` func called below.

Reply:
It is relevant because when creating a selector, only the parts of the state that need to be used in the transform function need to be isolated and obtained via an "input-selector". If all of the values returned from the input selectors are the same, then the memorized value can be returned. This avoids recomputing the return value of the transform function. See [docs about this here](https://www.npmjs.com/package/reselect#creating-a-memoized-selector).